### PR TITLE
REFACTOR: Implement full normalization with `TermView`.

### PR DESCRIPTION
**Pull request by wsx-ucb**
I've compared the new approach with the old one and there are two comments labeled `CHECK` that highlight the differences. May need to double check those changes are valid.

### CODE REVIEW

File: base/src/main/java/org/aya/core/visitor/TermOps.java [hunk 1]
```diff
@@ -132,4 +143,105 @@ public Term post(Term term) {
...
+          if (def == null) yield con;
+          var unfolded = unfoldClauses(true, con.conArgs(), def.clauses);
+          yield unfolded != null ? unfolded.data() : con;
+        }
+        case CallTerm.Fn fn -> {
+          var def = fn.ref().core;
+          if (def == null) yield fn;
+          if (def.modifiers.contains(Modifier.Opaque)) yield fn;
+          yield def.body.fold(
+            lamBody -> lamBody.view().subst(buildSubst(def.telescope(), fn.args())).rename().normal(state).commit(),
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/TermOps.java:208]: Do we really need to `rename` here, given that our `subst` has capture-avoiding?

File: base/src/main/java/org/aya/core/visitor/TermOps.java [hunk 2]
```diff
@@ -132,4 +143,105 @@ public Term post(Term term) {
...
+            patBody -> {
+              var orderIndependent = def.modifiers.contains(Modifier.Overlap);
+              var unfolded = unfoldClauses(orderIndependent, fn.args(), patBody);
+              // CHECK: Do we need to normalize `unfolded` again?
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/TermOps.java:212]: Yes.

File: base/src/main/java/org/aya/core/visitor/TermOps.java (hunk 1 above)
Comment (ice1000) [reply base/src/main/java/org/aya/core/visitor/TermOps.java:208]: (note that my comment also applies to other `rename` calls)

File: base/src/main/java/org/aya/core/visitor/TermOps.java [hunk 3]
```diff
@@ -132,4 +143,105 @@ public Term post(Term term) {
...
+          } else {
+            var subst = buildSubst(fieldDef.fullTelescope(), access.args());
+            for (var field : fieldDef.structRef.core.fields) {
+              if (field == fieldDef) continue;
+              var fieldArgs = field.telescope().map(Term.Param::toArg);
+              var acc = new CallTerm.Access(access.of(), field.ref, access.structArgs(), fieldArgs);
+              subst.add(field.ref, IntroTerm.Lambda.make(field.telescope(), acc));
+            }
+            var unfolded = unfoldClauses(true, access.fieldArgs(), subst, fieldDef.clauses);
+            // CHECK: seems no need to substitute `unfolded` again with `subst`
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/TermOps.java:231]: No.

File: base/src/main/java/org/aya/core/visitor/TermView.java [hunk 4]
```diff
@@ -151,6 +152,10 @@ private Term traverse(Term term) {
     return subst.isEmpty() ? this : new TermOps.Subster(this, subst);
   }
 
+  default TermView normal(TyckState state) {
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/visitor/TermView.java:155]: `normalize`

**Review by ice1000 — COMMENTED** [review]
Hope my comments are helpful :0

File: base/src/main/java/org/aya/core/visitor/TermOps.java (hunk 1 above)
Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/visitor/TermOps.java:208]: Ahh, the renaming is being done at substitution now, right?

File: base/src/main/java/org/aya/core/visitor/TermOps.java (hunk 2 above)
Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/visitor/TermOps.java:212]: I think everything unfolded are normalized before returning, see [here](https://github.com/aya-prover/aya-dev/pull/369/files#diff-349172ace414a6ad8b2b2a7fd8c4b9f90a8becf213add88455a7d42f886e535dR172).

Comment (ice1000) [reply base/src/main/java/org/aya/core/visitor/TermOps.java:212]: Gotcha

File: base/src/main/java/org/aya/core/visitor/TermOps.java (hunk 1 above)
Comment (ice1000) [reply base/src/main/java/org/aya/core/visitor/TermOps.java:208]: Yaas

**Review by ice1000 — APPROVED** [review]
bors d=wsx-ucb

**Comment by ice1000** [comment]
Whnfer when?

**Comment by wsx-ucb** [comment]
I'll try to finish that in the following days. But we'll see if that needs to use the `TermView` API.

**Pull request merged**